[clang][cas] Create explicit IncludeTree::ModuleMap

benlangmuir opened the pull request:
Instead of putting the modulemap file in the include-tree filesystem and parsing it at build time, create a data structure that represents just the parts of the modulemap we need when building the module:

* Flags (explicit, system, framework, etc.)
* Exports (export *, export Foo)
* LinkLibraries (implicit framework autolink, link Foo)

Additionally, we add modular headers lazily by inserting known headers when we encounter an include-tree header that is part of a submodule (this is needed for missing #include diagnostics).

This removes the possibility of mismatches between header paths seen during modulemap parsing from the paths we embed due to #includes, for exmaple due to VFS virtual vs external paths.

---

Other changes
* IncludeTree filesystem merges from imported PCH/modules: this gets input by merging across PCH/module include-trees. While this could pull in more files than necessary if there are files that are only needed during building, it turns out to be much better in practice for avoiding extra files we don't want such as modulemap files. It cannot cause extra cache misses, since all the files needed to build a module would already be in the cache keys for inputs.
* Tests for VFS, framework autolinking, inferred framework modules
* Handled visibility-only imports accurately

rdar://104386617

benlangmuir committed:
[clang][cas] Create explicit IncludeTree::ModuleMap

Instead of putting the modulemap file in the include-tree filesystem and
parsing it at build time, create a data structure that represents just
the parts of the modulemap we need when building the module:

* Flags (explicit, system, framework, etc.)
* Exports (export *, export Foo)
* LinkLibraries (implicit framework autolink, link Foo)

Additionally, we add modular headers lazily by inserting known headers
when we encounter an include-tree header that is part of a submodule
(this is needed for missing #include diagnostics).

This removes the possibility of mismatches between header paths seen
during modulemap parsing from the paths we embed due to #includes, for
exmaple due to VFS virtual vs external paths.

benlangmuir committed:
[clang][cas] Move include-tree and cas-fs IDs out of Module into ASTContext NFC

Unlike the cache key string, these are not needed for multiple modules
at once, and this allows us to store the include-tree ID for a PCH,
which will be useful in the next commit.

benlangmuir committed:
[clang][cas] IncludeTree filesystem merges from imported PCH/modules

Instead of looking through the PCH or modules for input files and
loading them into the current FileManager while scanning, merge them
directly from the include-tree that was used to build that PCH and/or
module. This makes it easier to understand where files come from in the
include-tree, and eliminates certain issues where the scanner module may
include more files than are necessary, or include them with different
paths in the case of VFS/symlinks.  It also avoids needing to re-stat
these files if they are not used during scanning the importing TU. As
part of this, we teach IncludeTreeFilesystem to handle ./ stripping in
paths to match what ASTWriter does.

benlangmuir committed:
[clang][cas] Add tests for include-tree module map

* Private modules with fmodule-name
* Inferred framework modules
* Framework autolink
* VFS

benlangmuir committed:
[clang][cas] Correctly split out modules that are only imported for visibility

Modules imported for visibility only can happen when -fmodule-name
matches a modular header that is imported more than once so that we
trigger the correct re-exports but without trying to import a module.

benlangmuir committed:
[clang][cas] Fix include-tree with pch importing public/private modules

We were dropping the modular import of a private module if its public
module was imported via PCH, because we skipped parsing the private
modulemap during scanning.

benlangmuir commented:
@swift-ci please test

akyrtzi commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -510,6 +500,19 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
+  if (!MF->CASFileSystemRootID.empty()) {
+    auto &CAS = MDC.ScanInstance.getOrCreateObjectStore();
+    if (auto Err = CAS.parseID(MF->CASFileSystemRootID)
+                       .moveInto(MD.CASFileSystemRootID)) {
```
Could you add a comment why `CASFileSystemRootID` is resolved to `CASID` but `IncludeTreeID` is kept a string?

benlangmuir replied:
I will look into making `CASFileSystemRootID` a string as well in a separate patch.  It doesn't really add value to have a CASID in this API.

akyrtzi commented on clang/lib/CAS/IncludeTree.cpp:
```diff
@@ -552,7 +860,11 @@ cas::createIncludeTreeFileSystem(IncludeTreeRoot &Root) {
           return FilenameBlob.takeError();
-        StringRef Filename = FilenameBlob->getData();
+        SmallString<128> Filename(FilenameBlob->getData());
+        // Strip './' in the filename to match the behaviour of ASTWriter; we
```
Is this affecting whether filenames in diagnostics show "./" or not? For example:
```
$ clang -fsyntax-only t.c
In file included from t.c:1:
./t.h:1:2: warning: blah [-W#warnings]
```
Not that this is so important to preserve, I'm mainly curious.

benlangmuir replied:
It will not change the diagnostic paths -- at least not for normal source file diagnostics.  The path in the include-tree is recorded as `./t.c`, which is what is used in the source manager. Only the IncludeTreeFilesystem removes the `./`.  Now, because `status` returns the path as seen by the filesystem, there is a possibility it leaks out without the `./` *somewhere*, but I think normal diagnostics will not change.

I will look at adding a test for this, but if you're okay with it I'll do that as a separate patch.

akyrtzi commented on clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp:
```diff
@@ -545,13 +562,82 @@ IncludeTreeBuilder::finishIncludeTree(CompilerInstance &ScanInstance,
+    if (Error E = Root->getFileList().moveInto(Files))
+      return E;
+
+    Error E = Files->forEachFile([&](auto IF, auto Size) -> Error {
```
Have you considered changing `IncludeTree::FileList` to be a CAS graph instead? We could have a graph of `FileList`s created from the module dependencies, that we flatten inside `cas::createIncludeTreeFileSystem()`.

benlangmuir replied:
Interesting idea, so the advantage is fewer CAS references created and moving the cost of deduplicating the files to compile time instead of scan time.

I'll take a look 👍

akyrtzi replied:
"fewer CAS references created" and we are not copying the `FileList`s to larger `FileList`s.

benlangmuir replied:
Implemented in d67f7857d04c277.

benlangmuir commented:
@swift-ci please test

benlangmuir committed:
[clang][cas] Make missing include-tree id an error

This was review feedback on a previous PR that somehow went missing.
Bring it back.

benlangmuir commented:
@swift-ci please test

akyrtzi commented on clang/lib/CAS/IncludeTree.cpp:
```diff
@@ -268,6 +278,236 @@ bool IncludeTree::FileList::isValid(const ObjectProxy &Node) {
          Base.getData().size() == NumFiles * sizeof(FileSizeTy);
 }
 
+static constexpr char ModuleFlagFramework = 0x01;
```
I'd prefer to write it like this for improved readability:
```
static constexpr char ModuleFlagFramework = 1 << 0;
static constexpr char ModuleFlagExplicit = 1 << 1;
static constexpr char ModuleFlagExternC = 1 << 2;
static constexpr char ModuleFlagSystem = 1 << 3;
static constexpr char ModuleFlagHasExports = 1 << 4;
static constexpr char ModuleFlagHasLinkLibraries = 1 << 5;
```

benlangmuir replied:
Fixed in latest push.

akyrtzi commented on clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp:
```diff
@@ -545,13 +565,82 @@ IncludeTreeBuilder::finishIncludeTree(CompilerInstance &ScanInstance,
+      if (Error E = AddModule(M))
+        return std::move(E);
+      Module *PM =
+          MMap.findModule(ScanInstance.getLangOpts().ModuleName + "_Private");
```
Is the "_Private" suffix just a convention? Is it possible some framework picks a different suffix for the private module and we fail to find it here?

benlangmuir replied:
It's handled specifically, not just a convention. There are various places in ModuleMap that handle private modules by name.  Technically there are three kinds (Foo_Private, Foo.Private, and FooPrivate), but only Foo_Private is relevant here, because that's the only one that is (a) treated as part of the main module in the function `Module::isForBuilding`, and (b) is not already a submodule that would be included as part of `AddModule(M)` above.

akyrtzi commented on clang/lib/Lex/HeaderSearch.cpp:
```diff
@@ -1773,8 +1773,11 @@ HeaderSearch::lookupModuleMapFile(const DirectoryEntry *Dir, bool IsFramework) {
                                           bool IsSystem) {
-  if (Module *Module = ModMap.findModule(Name))
-    return Module;
+  // CAS: we need to load modulemaps even if we have the primary module for this
```
@jansvoboda11 do you have any concerns with this change?

benlangmuir replied:
To clarify, when we upstream CAS/caching, my intention would be to just delete this check.  I'm only keeping this comment for the benefit of merge conflict handling.

benlangmuir replied:
I found a case where this causes "error: redefinition of module". Looking into what else can work here.

benlangmuir replied:
This code is removed. I am instead triggering module lookups during scanning a PCH that should behave the same as the module lookups we would do when reading modules without include-tree, and cause us to lookup the modulemap files we need.

akyrtzi reviewed (commented):
General comment, we are adding more and more tests related to cas/caching, should we move them under `test/ClangScanDeps/CAS` for better organization?

benlangmuir committed:
[clang][cas] Add simple prefix-mapping test for include-tree modules

benlangmuir committed:
[clang][cas] Allow non-existent fmodule-name

Missing the module named by fmodule-name is not an error for a TU or
PCH.

benlangmuir committed:
[clang][cas] Switch to chained representation of IncludeTree::FileList

Instead of flattening file lists when importing PCH/modules, just create
a chained reference to the list and flatten it only in
FileList::forEachFile. This reduces the storage cost and performance
overhead of merging file lists.

benlangmuir commented:
@swift-ci please test

akyrtzi commented on clang/lib/CAS/IncludeTree.cpp:
```diff
@@ -222,29 +240,44 @@ IncludeTree::FileList::getFileSize(size_t I) const {
+    if (!Proxy)
+      return Proxy.takeError();
+    FileList FL(std::move(*Proxy));
+    return FL.forEachFile(Callback);
```
Could you enhance `FileList::forEachFile()` and have it skip a `FileList` that was already visited?

benlangmuir replied:
SGTM. Also easy to skip visiting individual files that are identical and contained in multiple lists.

benlangmuir replied:
Fixed in 2066527eca2d

benlangmuir committed:
[clang][cas] Skip duplicates in forEachFile

If we see the same File or FileList more than once, skip visiting it
again. This improves performance if the same module is reached through
multiple dependency chains, or the same file is in multiple
modules/PCH/TU file lists explicitly.

benlangmuir commented:
@swift-ci please test

akyrtzi reviewed (approved):
Great work! 🎉
